libpoolprotocols/PoolManager.cpp

```diff
@@ -9,247 +10,223 @@ PoolManager::PoolManager(boost::asio::io_service& io_service, PoolClient* client
+                cwarn << "Stale solution: " << EthWhite "0x" << toHex(sol.nonce) << EthReset;
+            else
+                cnote << "Solution: " << EthWhite "0x" << toHex(sol.nonce) << EthReset;
+
+            p_client->submitSolution(sol);
+        }
+        else
+        {
+            cnote << string(EthRed "Solution 0x") + toHex(sol.nonce)
+                  << " wasted. Waiting for connection ...";
+        }
+
+        return false;
+    });
+    m_farm.onMinerRestart([&]() {
+        dev::setThreadName("main");
+        cnote << "Restart miners...";
+
+        if (m_farm.isMining())
+        {
+            cnote << "Shutting down miners...";
+            m_farm.stop();
+        }
+
+        cnote << "Spinning up miners...";
+        if (m_minerType == MinerType::CL)
+            m_farm.start("opencl", false);
+        else if (m_minerType == MinerType::CUDA)
+            m_farm.start("cuda", false);
+        else if (m_minerType == MinerType::Mixed)
+        {
+            m_farm.start("cuda", false);
+            m_farm.start("opencl", true);
+        }
+    });
 }
 
 void PoolManager::stop()
 {
-	if (m_running.load(std::memory_order_relaxed)) {
-		
-		cnote << "Shutting down...";
-
-		m_running.store(false, std::memory_order_relaxed);
-		m_failovertimer.cancel();
-
-		if (p_client->isConnected()) 
-			p_client->disconnect(); 
-
-		if (m_farm.isMining())
-		{
-			cnote << "Shutting down miners...";
-			m_farm.stop();
-		}
-
-	}
+    if (m_running.load(std::memory_order_relaxed))
+    {
+        cnote << "Shutting down...";
+
+        m_running.store(false, std::memory_order_relaxed);
+        m_failovertimer.cancel();
+
+        if (p_client->isConnected())
+            p_client->disconnect();
+
+        if (m_farm.isMining())
+        {
+            cnote << "Shutting down miners...";
+            m_farm.stop();
+        }
+    }
 }
 
 void PoolManager::workLoop()
 {
-
-	dev::setThreadName("main");
-
-	while (m_running.load(std::memory_order_relaxed))
-	{
-
-		// Take action only if not pending state (connecting/disconnecting)
-		// Otherwise do nothing and wait until connection state is NOT pending
-		if (!p_client->isPendingState()) {
-
-			if (!p_client->isConnected()) {
-
-				// If this connection is marked Unrecoverable then discard it
-				if (m_connections.at(m_activeConnectionIdx).IsUnrecoverable())
-				{
+    dev::setThreadName("main");
+
+    while (m_running.load(std::memory_order_relaxed))
+    {
+        // Take action only if not pending state (connecting/disconnecting)
+        // Otherwise do nothing and wait until connection state is NOT pending
+        if (!p_client->isPendingState())
+        {
+            if (!p_client->isConnected())
+            {
+                // If this connection is marked Unrecoverable then discard it
+                if (m_connections.at(m_activeConnectionIdx).IsUnrecoverable())
+                {
                     p_client->unsetConnection();
 
-                    /*
```

Comment:
Instead of creating a new vector, do an in-place remove

Reply:
Already tried.
As I wrote in the comment method erase on vector always brings to unspecified behavior: subsequent calls to remaining elements addresses to unspecified memory location

Reply:
There is an underlying problem here... remove_if definitely works with vectors! Must be something to do with the URI copy constructor. Let me dig a little.